### DISCUSSION TITLE
Plugin-ise output metadata modules

### DISCUSSION THREAD

**Discussion by mizaki** [comictagger/comictagger#481] [Ideas]
This should be the last thing to be plugin-ised. 

I think it should be pretty simple outside of collisions. I know CBI/CBL writes to the archive comment but does anything else? If so, can you store both or should you? Other than that just making sure two plugins aren't trying to write to the same file and overwriting each other.

The class itself probably only needs `write_metadata` and `read_metadata` and any options that might be wanted (build it in the GUI the same as talkers).

The reason for options means that more "advanced" output plugins could be created such as one that will HTTP POST or even one that could write to a SQL DB (or SQLite) etc.

The other new ability would be running multiple output plugins. That would mean a GUI change. It could be moved to the settings menu with the output settings having a required options of `active/inactive` but should it also be represented on the main window?

Any other ideas/thoughts to go with it?

**Comment by lordwelch** [comment]
Are you talking about metadata plugins?

**Reply by mizaki** [reply]
Yep. Updated title.

I also realised there will probably need some work on the writing of files via `comicarchive` and writing files in general.

**Comment by mizaki** [comment]
Writing to multiple file etc. should not be a problem but what about reading?

1. Use something similar to now and allow only one external source of metadata?
2. Use an overlay and combine multiple (this is already in place for metadata in general?)?

With two, it would be how to set the reading order. GUI could use a sorted list and active/inactive for reading. On the CLI, take the order they are entered?

**Reply by lordwelch** [reply]
We already have everything in place for reading multiple types of metadata, the CLI has 3 and the GUI only has 2 (I'm not sure why CoMet wasn't added to the GUI). I don't see a reason to change how it works at the moment.

I already have code that does this I just need to fix my laptop that has some hardware issues at the moment...

**Reply by mizaki** [reply]
I haven't check in the code yet but there is no overlay in the GUI. If I save say have a title in CBL and switch to read style CIX it will disappear. That's what I mean with 1. Even the CLI help states:
```
-t {CR,CBL,COMET}, --type {CR,CBL,COMET}
                        Specify TYPE as either CR, CBL or COMET
                        (as either ComicRack, ComicBookLover,
                        or CoMet style tags, respectively).
                        Use commas for multiple types.
                        For searching the metadata will use the first listed:
                        eg '-t cbl,cr' with no CBL tags, CR will be used if they exist
```
In case I wasn't clear with 2. I mean if there is series in CIX, issue number in CBL and year in COMET then overlay all three so search can use series, issue number and year.

**Reply by lordwelch** [reply]
That is an entirely different feature request. If you have an idea for merging info from multiple metadata styles or otherwise changing how ComicTagger deals with multiple metadata styles you can make a pull request for that now, it has little bearing on turning it into a plugin.

**Reply by mizaki** [reply]
I was thinking ahead (as I like to do). The initial process would be the same as for talkers. Convert what there is now and go from there. I'll start another discussion on the metadata.
